Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2Connections.swift

```diff
@@ -0,0 +1,532 @@
+            case .starting:
+                return .keepConnection
+            case .backingOff:
+                context.connectBackoff.append(self.connectionID)
+                return .removeConnection
+            case .active(let connection, _, let usedStreams, _):
+                if usedStreams <= 0 {
+                    context.close.append(connection)
+                    return .removeConnection
+                } else {
+                    context.cancel.append(connection)
+                    return .keepConnection
+                }
+            case .draining(let connection, _, _):
+                context.cancel.append(connection)
+                return .keepConnection
+            case .closed:
+                preconditionFailure("Unexpected state for cleanup: Did not expect to have closed connections in the state machine.")
+            }
+        }
+        
+        func addStats(into stats: inout HTTP2Connections.Stats) {
+            if self.isIdle {
+                stats.idleConnections &+= 1
+            }
+            switch self.state {
+            case .starting:
+                stats.startingConnections &+= 1
+            case .backingOff:
+                stats.backingOffConnections &+= 1
+            case .active(_, let maxStreams, let usedStreams, _):
+                stats.availableStreams += max(maxStreams - usedStreams, 0)
+                stats.leasedStreams += usedStreams
+                stats.availableConnections &+= 1
+            case .draining(_, _, let usedStreams):
+                stats.drainingConnections &+= 1
+                stats.leasedStreams += usedStreams
+            case .closed:
+                break
+            }
+        }
+
+        init(connectionID: Connection.ID, eventLoop: EventLoop) {
+            self.connectionID = connectionID
+            self.eventLoop = eventLoop
+            self.state = .starting
+        }
+    }
+
+    struct HTTP2Connections {
+        /// Information around the failed/closed connection.
+        struct FailedConnectionContext {
+            /// The eventLoop the connection ran on.
+            var eventLoop: EventLoop
+        }
+
+        /// A connectionID generator.
+        private let generator: Connection.ID.Generator
+        /// The connections states
+        private var connections: [HTTP2ConnectionState]
+        
+        var isEmpty: Bool {
+            connections.isEmpty
+        }
+        
+        var stats: Stats {
+            connections.reduce(into: Stats()) { stats, connection in
+                connection.addStats(into: &stats)
+            }
+        }
+
+        init(generator: Connection.ID.Generator) {
+            self.generator = generator
+            self.connections = []
+        }
+        
+        // MARK: Migration
+        
+        mutating func applyMigration(_ migration: HTTP1Connections.MigrationContext) {
+            for (connectionID, eventLoop) in migration.starting {
```

Comment:
add comment here, that in this struct we can only handle starting connections. all running connections must be transitioned which is the job of the enclosing state machine.

Reply:
Added a comment. I also do no longer give this function the full `MigrationContext` and instead only pass `starting` and `backingOff` connections to make it even more clear.